### ISSUE TITLE
Initializing IEnumberable<T> that implements Iterable<T>

### ISSUE THREAD

**Issue by hu13** [saltosystems/winrt-go#100]
hi there.

i am trying to use this api
https://learn.microsoft.com/en-us/uwp/api/windows.storage.provider.storageprovideritemproperties.setasync?view=winrt-26100#windows-storage-provider-storageprovideritemproperties-setasync(windows-storage-istorageitem-windows-foundation-collections-iiterable((windows-storage-provider-storageprovideritemproperty)))
```
public static IAsyncAction SetAsync(IStorageItem item, IEnumerable<StorageProviderItemProperty> itemProperties);
```
and it requires initializing a list of `StorageProviderItemProperty` that implements Iterable<StorageProviderItemProperty>. I looked at the interfaces that go generate came up for
```
//go:generate go run github.com/saltosystems/winrt-go/cmd/winrt-go-gen -debug -class Windows.Foundation.Collections.IIterable`1
//go:generate go run github.com/saltosystems/winrt-go/cmd/winrt-go-gen -debug -class Windows.Foundation.Collections.IIterator`1
```
but i can't seem to see a pattern on how to generate a `Iterable<StorageProviderItemProperty>` from that. I am wondering if you have any experience interacting with Iterable interface from winrt-go before and can provide some insight. Thanks.

apologize if this is a bit off topics.

**Comment by jagobagascon** [comment]
We've worked with buffers and vectors, but always as readers, we never had to deal with creating an iterable.

And it seems that [there is no specific class](https://learn.microsoft.com/en-us/uwp/api/windows.foundation.collections?view=winrt-26100#classes) in `Windows.Foundation.Collections` that implements the `IIterable` interface. So the only solution I can think of is to provide your own implementation.

I will try to look into it, but I'll have to find some free time to do so because I think this is going to be quite complex to do.

In the meantime you can check how the delegates do this by defining their own functions `VTable` using `syscall.NewCallback`.

**Comment by jagobagascon** [comment]
I wanted to help a little bit more so I made a quick test that seems to be working. I've uploaded it as a gist: https://gist.github.com/jagobagascon/fd807800ee5c658955a81f0dd1c40bb8

You should be able to run it after generating the iterable and iterator structs.

It's far from perfect and required quite a bit of black magic to make it work:
```go
func moveNext(inst, out unsafe.Pointer) uintptr {
	offset := unsafe.Offsetof(collectionsIterator{}.IIterator)
        // good old C pointer magic tricks
	it := (*collectionsIterator)(unsafe.Pointer(uintptr(inst) - offset))

	it.index++
	return getHasCurrent(inst, out)
}
```

I hope this helps.

> [!NOTE]
> I don't think we will have time to work on this in the short term, so PRs are welcome 😉

**Comment by jagobagascon** [comment]
You are right, sorry. I refactored the delegates and extracted the code shared among all structs to the `iunknown` package. That way you won't need to implement `Invoke` (which is only required by delegates).

I pushed the changes to this branch: https://github.com/saltosystems/winrt-go/tree/feature/slice-iterable-testing-prototype

With them the `main.go` added to the Gist should work as expected.

**Comment by jagobagascon** [comment]
I fixed the item assignment using reflection: https://gist.github.com/jagobagascon/fd807800ee5c658955a81f0dd1c40bb8#file-array-go-L356
